### ISSUE TITLE
Run tests with pytest directly instead of via tests/run_all_tests.sh

### ISSUE THREAD

**Issue by danielsuo** [google/flax#230]
After some discussion with @avital, I started a PR, but saw that the change may be a bit more involved. Opening an issue here for feedback.

# Introduction
**Where we are today**: We use `tests/run_all_tests.sh` because our `example` tests have conflicting names (`train_test.py`) and there are some additional checks (e.g., warn if example doesn't have tests). Perhaps by design, `tests/run_all_tests.sh` also lets us have very self-contained examples by collecting/running each one with different `pytest` calls.

**Why that might be suboptimal**: Testing can quickly become brittle. As an example, I wanted to avoid a second call to `pytest tests` while creating coverage reports, so I made [these changes](https://github.com/google/flax/pull/227/files#diff-bcf3e47075e6daafc409adc7a3c0787a) to `tests/run_all_tests.sh`. Not my best work...

**What's the blocker**: `absl` maintains a global `FLAGS` object that raises a `DuplicateFlagError` exception because multiple examples use, for instance, `learning_rate` as a flag. Because `pytest` collects tests before distributing to `pytest-xdist`, we fail.

**Possible solution**: @avital mentioned that @Marvin182 suggested splitting

> all examples into, e.g., `mnist_train.py` (just a train function, no flags, no main) and `mnist_main.py` (with `FLAGS`).

At the end of the day, hopefully this minimally impacts both example maintainers (i.e., follow some documentation to set up an example in the expected way) and users (can still `cd` in and run).

# Proposal
- Standardize example format into `{example}_train.py`, `{example}_main.py`, and any supplementary files, as noted above
- Turn each example into a module (i.e., add `__init__.py`) and switch to relative imports

# Transition
- One PR / example with reviews from example maintainers
- Add notes to `flax/examples/README.md` describing the new example structure
- Remove `tests/run_all_tests.sh`. Given the philosophy re examples, I don't think we need an automated check. If we want all examples to have tests, we just make them have tests.

**Comment by Marvin182** [comment]
I just created #231 to add guidelines for examples. As part of this I'm also advocating to split examples into a library without flags and a small main file with the flags.

I'm not sure I agree with the idea to add a `__init__.py` and switch to relative imports. What benefit would that provide?

**Comment by danielsuo** [comment]
I believe `pytest` will complain without since `examples` is not a package and `pytest` is running from the root repository directory, but I haven't looked very carefully into this. Do you have suggestions?

**Comment by danielsuo** [comment]
Believed I figured it out; corrected the proposal.

**Comment by avital** [comment]
I agree that we should wait for #231 after which this should be basically automatically resolved. Keeping the issue open to remind us to update our testing documentation once that's in.

**Comment by danielsuo** [comment]
Yeah, this will be very helpful so that we can trivially increase parallelism as the number of examples grows.

**Comment by avital** [comment]
Looks like in #231 it's been decided that not all examples must use exactly the same directory structure, so it seems unlikely that we'll be able to guarantee that no two test directories contains the same filenames. So I don't know if we'll be able to resolve this. I'll mark it closed, but if people have thoughts on this please let us know!
